





















# Hospital Readmission Prediction

## Notebook 01 - Data Understanding

**Author:** Ketharnath R

### Objective
The objective of this notebook is to understand the dataset before performing preprocessing or building machine learning models.

In [5]:
import pandas as pd
import numpy as np

In [6]:
df = pd.read_csv("../data/raw/hospital_readmissions.csv")

In [7]:
df.head()

,age,time_in_hospital,n_lab_procedures,n_procedures,n_medications,n_outpatient,n_inpatient,n_emergency,medical_specialty,diag_1,diag_2,diag_3,glucose_test,A1Ctest,change,diabetes_med,readmitted
0,[70-80),8,72,1,18,2,0,0,Missing,Circulatory,Respiratory,Other,no,no,no,yes,no
1,[70-80),3,34,2,13,0,0,0,Other,Other,Other,Other,no,no,no,yes,no
2,[50-60),5,45,0,18,0,0,0,Missing,Circulatory,Circulatory,Circulatory,no,no,yes,yes,yes
3,[70-80),2,36,0,12,1,0,0,Missing,Circulatory,Other,Diabetes,no,no,yes,yes,yes
4,[60-70),1,42,0,7,0,0,0,InternalMedicine,Other,Circulatory,Respiratory,no,no,no,yes,no


In [8]:
df.shape

(25000, 17)

In [9]:
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   age                25000 non-null  str  
 1   time_in_hospital   25000 non-null  int64
 2   n_lab_procedures   25000 non-null  int64
 3   n_procedures       25000 non-null  int64
 4   n_medications      25000 non-null  int64
 5   n_outpatient       25000 non-null  int64
 6   n_inpatient        25000 non-null  int64
 7   n_emergency        25000 non-null  int64
 8   medical_specialty  25000 non-null  str  
 9   diag_1             25000 non-null  str  
 10  diag_2             25000 non-null  str  
 11  diag_3             25000 non-null  str  
 12  glucose_test       25000 non-null  str  
 13  A1Ctest            25000 non-null  str  
 14  change             25000 non-null  str  
 15  diabetes_med       25000 non-null  str  
 16  readmitted         25000 non-null  str  
dtypes: int64(7), str(10)
me

## Observation

The dataset contains **25,000 patient records** and **17 features**.

The target variable is **readmitted**, which makes this a **binary classification** problem.

No preprocessing has been performed yet.

----------------

# Step 6: Statistical Summary

The `describe()` function provides descriptive statistics for all numerical features.

It includes:
- Count
- Mean
- Standard Deviation
- Minimum
- Maximum
- Quartiles (25%, 50%, 75%)

These statistics help us understand the distribution of numerical data and identify unusual values.

In [10]:
df.describe()

,time_in_hospital,n_lab_procedures,n_procedures,n_medications,n_outpatient,n_inpatient,n_emergency
count,25000.00000,25000.00000,25000.000000,25000.000000,25000.000000,25000.000000,25000.000000
mean,4.45332,43.24076,1.352360,16.252400,0.366400,0.615960,0.186600
std,3.00147,19.81862,1.715179,8.060532,1.195478,1.177951,0.885873
min,1.00000,1.00000,0.000000,1.000000,0.000000,0.000000,0.000000
25%,2.00000,31.00000,0.000000,11.000000,0.000000,0.000000,0.000000
50%,4.00000,44.00000,1.000000,15.000000,0.000000,0.000000,0.000000
75%,6.00000,57.00000,2.000000,20.000000,0.000000,1.000000,0.000000
max,14.00000,113.00000,6.000000,79.000000,33.000000,15.000000,64.000000


## 🔍 Observations

Based on the statistical summary (`df.describe()`), the following observations can be made:

- The dataset contains **25,000 patient records** for all numerical features, indicating that there are no missing values in these columns.

- The average **hospital stay** is approximately **4.45 days**, with stays ranging from **1 day** to **14 days**.

- On average, a patient undergoes approximately **43 laboratory procedures** during hospitalization. The number of lab procedures varies between **1 and 113**.

- Patients undergo an average of **1.35 medical procedures**, with a maximum of **6 procedures**.

- The average number of **medications prescribed** is approximately **16**, while some patients receive as many as **79 medications**.

- Most patients have **no outpatient, inpatient, or emergency visits** before admission, as indicated by the median (50th percentile) values of **0**.

- However, a few patients have a significantly higher number of previous visits:
  - Outpatient visits: up to **33**
  - Inpatient visits: up to **15**
  - Emergency visits: up to **64**

- The large difference between the median and maximum values for previous hospital visits suggests the presence of **outliers**. These patients may represent individuals with chronic illnesses or frequent hospital admissions.

## 📌 Conclusion

The numerical features show considerable variation among patients. Some variables, particularly the number of medications and previous hospital visits, contain extreme values that may require further investigation during Exploratory Data Analysis (EDA).

In the next phase, we will visualize these features using charts and identify distributions, trends, and potential outliers before preprocessing the data.

-----------


# Step 7: Feature Names

## Objective

The `columns` attribute returns the names of all columns (features) present in the dataset.

Understanding the available features is important because each column represents a specific patient attribute that may influence hospital readmission.

By examining the feature names, we can:

- Identify the input features (independent variables).
- Identify the target variable (dependent variable).
- Understand what type of information is available for analysis and prediction.

In [11]:
df.columns


Index(['age', 'time_in_hospital', 'n_lab_procedures', 'n_procedures',
       'n_medications', 'n_outpatient', 'n_inpatient', 'n_emergency',
       'medical_specialty', 'diag_1', 'diag_2', 'diag_3', 'glucose_test',
       'A1Ctest', 'change', 'diabetes_med', 'readmitted'],
      dtype='str')

## 🔍 Observations

The dataset contains **17 columns**, representing different characteristics of patients and their hospital visits.

Some important features include:

- **age** – Age group of the patient.
- **time_in_hospital** – Number of days the patient stayed in the hospital.
- **n_lab_procedures** – Number of laboratory procedures performed.
- **n_procedures** – Number of medical procedures performed.
- **n_medications** – Number of medications prescribed.
- **medical_specialty** – Medical department responsible for the patient's treatment.
- **diag_1** – Primary diagnosis of the patient.
- **glucose_test** – Result of the glucose test.
- **A1Ctest** – Result of the HbA1c test.
- **change** – Indicates whether there was a change in medication during hospitalization.
- **diabetes_med** – Indicates whether the patient was prescribed diabetes medication.
- **readmitted** – Indicates whether the patient was readmitted after discharge.

## 📌 Conclusion

The dataset contains a combination of demographic information, medical history, laboratory results, diagnosis details, and treatment information.

The **`readmitted`** column is the **target variable**, while all other columns serve as input features that will be used to train the machine learning model.

----------

# Step 8: Data Types

## Objective

The `dtypes` attribute displays the data type of each feature in the dataset.

Understanding the data types is an essential step before preprocessing because machine learning algorithms can only work with numerical data.

The common data types in this dataset include:

- **int64** – Integer values (whole numbers)
- **object** – Text or categorical values

By identifying the data types, we can determine which features require preprocessing, such as encoding categorical variables into numerical values.

In [12]:
df.dtypes

age                    str
time_in_hospital     int64
n_lab_procedures     int64
n_procedures         int64
n_medications        int64
n_outpatient         int64
n_inpatient          int64
n_emergency          int64
medical_specialty      str
diag_1                 str
diag_2                 str
diag_3                 str
glucose_test           str
A1Ctest                str
change                 str
diabetes_med           str
readmitted             str
dtype: object

## 🔍 Observations

The dataset contains both **numerical** and **categorical** features.

### Numerical Features (int64)

These columns contain numerical values and can be directly used for statistical analysis and machine learning.

Examples include:

- time_in_hospital
- n_lab_procedures
- n_procedures
- n_medications
- n_outpatient
- n_inpatient
- n_emergency

### Categorical Features (object)

These columns contain text-based or category-based information.

Examples include:

- age
- medical_specialty
- diag_1
- diag_2
- diag_3
- glucose_test
- A1Ctest
- change
- diabetes_med
- readmitted

Since machine learning algorithms cannot directly process text values, these categorical features must be converted into numerical values during the preprocessing stage using encoding techniques.

## 📌 Conclusion

The dataset consists of both numerical and categorical features.

The numerical features are already suitable for analysis, while the categorical features will require preprocessing before model training.

The target variable **`readmitted`** is also categorical and will need to be encoded into numerical values before building the machine learning model.

-----------------

---
# 📖 Step 9: Missing Values
---

## 🎯 Objective

In real-world datasets, some values may be missing due to incomplete records, data entry errors, or unavailable information.

The `isnull().sum()` function is used to identify the number of missing values present in each feature.

Identifying missing values is an important preprocessing step because most machine learning algorithms cannot handle missing data directly.

By analyzing missing values, we can:

- Determine whether the dataset contains incomplete information.
- Identify which features require data cleaning.
- Decide the appropriate strategy for handling missing values, such as removing records or imputing values.

In [13]:
df.isnull().sum()

age                  0
time_in_hospital     0
n_lab_procedures     0
n_procedures         0
n_medications        0
n_outpatient         0
n_inpatient          0
n_emergency          0
medical_specialty    0
diag_1               0
diag_2               0
diag_3               0
glucose_test         0
A1Ctest              0
change               0
diabetes_med         0
readmitted           0
dtype: int64

## 🔍 Observations

Analyze the output of `df.isnull().sum()` and note the following observations.

- Check whether any feature contains missing values.
- Identify the columns with the highest number of missing values.
- Determine whether the missing values are present in numerical or categorical features.
- Consider whether the amount of missing data is significant enough to require preprocessing.

If every column has a value of **0**, it indicates that no missing values are detected by Pandas in the dataset.

If one or more columns contain values greater than **0**, those features will require appropriate handling during the preprocessing stage.

## 📌 Conclusion

Understanding missing values is a critical step before building a machine learning model.

The results of this analysis will guide the preprocessing techniques used later in the project, ensuring that the dataset is clean and suitable for model training.



---
# 📖 Step 10: Target Variable Analysis
---

## 🎯 Objective

The target variable is the output that a machine learning model is trained to predict.

In this project, the target variable is **`readmitted`**, which indicates whether a patient was readmitted to the hospital after discharge.

Before building a machine learning model, it is important to understand the distribution of the target variable.

Analyzing the target variable helps us:

- Identify the different classes present in the dataset.
- Determine whether the dataset is balanced or imbalanced.
- Understand the proportion of patients who were readmitted.
- Select appropriate evaluation metrics and preprocessing techniques.

In [14]:
df["readmitted"].value_counts()

readmitted
no     13246
yes    11754
Name: count, dtype: int64

## 🔍 Observations

The target variable **`readmitted`** consists of two classes:

- **No:** 13,246 patients were **not readmitted** after discharge.
- **Yes:** 11,754 patients were **readmitted** after discharge.

The difference between the two classes is relatively small, indicating that the dataset is **fairly balanced**.

A balanced dataset is beneficial because the machine learning model can learn patterns from both classes without being heavily biased toward one class.

However, the exact class proportions should also be examined using percentage distribution to confirm the balance.

## 📌 Conclusion

The target variable represents a **binary classification problem**, where the objective is to predict whether a patient will be **readmitted** (`yes`) or **not readmitted** (`no`).

Since both classes are present in comparable numbers, this dataset is well-suited for training classification models without requiring major class balancing techniques.

-------------

## 📊 Percentage Distribution

To better understand the class distribution, we calculate the percentage of each class.

In [16]:
(df["readmitted"].value_counts(normalize=True) * 100).round(2)

readmitted
no     52.98
yes    47.02
Name: proportion, dtype: float64

## 🔍 Observations

The percentage distribution of the target variable is as follows:

- **No:** **52.98%** of patients were **not readmitted** after discharge.
- **Yes:** **47.02%** of patients were **readmitted** after discharge.

The difference between the two classes is relatively small (approximately **6%**), indicating that the dataset is **fairly balanced**.

A balanced dataset is advantageous because the machine learning model has sufficient examples from both classes to learn meaningful patterns. As a result, the model is less likely to become biased toward one particular class.

## 📌 Conclusion

The `readmitted` feature is the **target variable** of this project, making it a **binary classification problem** with two possible outcomes: **Yes** and **No**.

Since the class distribution is nearly balanced (52.98% vs. 47.02%), the dataset is suitable for training classification models without requiring major class-balancing techniques. However, model performance will still be evaluated using appropriate classification metrics such as **Accuracy, Precision, Recall, F1-Score, and ROC-AUC** to ensure reliable predictions.

----------------------

---
# 📋 Dataset Summary
---

## Key Findings

After understanding the dataset, the following observations were made:

- The dataset contains **25,000 patient records**.
- There are **17 features**, including the target variable.
- The dataset consists of both **numerical** and **categorical** features.
- No missing values were detected.
- The target variable (`readmitted`) represents a binary classification problem.
- The class distribution is nearly balanced (52.98% vs. 47.02%).
- Numerical features show varying ranges and may require scaling during preprocessing.
- Categorical features will require encoding before model training.

## Overall Conclusion

The dataset is clean and suitable for further analysis. The next phase is Exploratory Data Analysis (EDA), where visualizations will be used to identify patterns, trends, and relationships among the features.